daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala

```diff
@@ -48,6 +51,72 @@ object XMLTextInfosetInputter {
     fact.setProperty(XMLInputFactory.IS_VALIDATING, false) // no DTD validation
     fact
   }
+
+  lazy val xmlOutputFactory = {
+    val fact = new com.ctc.wstx.stax.WstxOutputFactory()
+    fact
+  }
+
+  /**
+   * When the stringAsXml runtime property is set to true we treat the simple
+   * content as if it were XML and output it to the infoset. We use this value
+   * as the runtime property name. Additionally, when we output the XML we need
+   * a wrapper element to reset the default XML namespace in case the infoset
+   * has a default namespace. We also use this variable as the name of that
+   * wrapper element.
+   */
+  lazy val stringAsXml = "stringAsXml"
+
+  /**
+   * Read an event from an XMLStreamReader and write the same event to an
+   * XMLStreamWriter. For example, this can be used for convert XML from one
+   * type of input to a different type (e.g. InputStream to String), check for
+   * well-formed XML, and perform some basic normalization. Example usage of
+   * this function is to call this in a loop like this:
+   *
+   *   while (xsr.hasNext()) {
+   *     XMLTextInfoset.writeStreamEvent(xsr, xsw)
+   *     xsr.next()
+   *   }
+   *
+   * This approach allows callers to skip certain events by conditionally
+   * calling this function.
+   */
+  def writeXMLStreamEvent(xsr: XMLStreamReader, xsw: XMLStreamWriter): Unit = {
+    xsr.getEventType() match {
+      case START_ELEMENT => {
+        xsw.writeStartElement(
+          xsr.getPrefix(),
+          xsr.getLocalName(),
+          xsr.getNamespaceURI())
+        for (i <- 0 until xsr.getNamespaceCount()) {
+          xsw.writeNamespace(
+            xsr.getNamespacePrefix(i),
+            xsr.getNamespaceURI(i))
+        }
+        for (i <- 0 until xsr.getAttributeCount()) {
+          xsw.writeAttribute(
+            xsr.getAttributePrefix(i),
+            xsr.getAttributeNamespace(i),
+            xsr.getAttributeLocalName(i),
+            xsr.getAttributeValue(i))
+        }
+      }
+      case END_ELEMENT => xsw.writeEndElement()
+      case CHARACTERS => xsw.writeCharacters(xsr.getText())
+      case COMMENT => xsw.writeComment(xsr.getText())
```

Comment:
So we are able to preserve comments and processing instructions. That's pretty interesting. I wonder if whitespace within those will be preserved or collapsed/modified. Our test data embedded XML string should test this to see. More on this in later comment about test enhancement. 

I see below even entity references seem like they're being preserved. 

So is this why the coalescing 'false' setting is used?

Of interesting note, if we created a corresponding feature for the EXI inputter/outputter, it would NOT be able to preserve many of these things, as EXI itself doesn't preserve many of them.

Reply:
Yep, I *think* much of this will be preserved. Though I didn't test it in much detail. This is really just a proxy function that converts XMLStreamReader events to XMLStreamWriter events. It's pretty straightforward to just handle all the events. I'm not sure exactly in what circumstances the woodstox XMLStreamReader will actually use those events. For example, with coalescing turned off, i noticed CDATA events were never created from the reader. But if it ever does, we can support them.